README.md
#### `master` [![CircleCI](https://circleci.com/gh/PMA-2020/pma-api/tree/master.svg?style=svg&circle-token=3cd5fffe3dad1d27e6cc1000307bc299e2ef3e77)](https://circleci.com/gh/PMA-2020/pma-api/tree/master) |  `develop` [![CircleCI](https://circleci.com/gh/PMA-2020/pma-api/tree/develop.svg?style=svg&circle-token=3cd5fffe3dad1d27e6cc1000307bc299e2ef3e77)](https://circleci.com/gh/PMA-2020/pma-api/tree/develop)

# About the PMA API
This is the PMA2020 API. You can access it at: http://api.pma2020.org

# 1. User Documentation
User documentation is accessible at: http://api-docs.pma2020.org

# 2. Developer Documentation
## 2.1. Using the API
### Navigation
#### Resource Lists
Resource roots are endpoints that return a list of resources and URLs to those
resources. The resource list for common endpoints can be found at `/v#` or
`/v#/resources`, and the resource list for application-specific endpoints can
be found at `/v#/datalab` or `v#/datalab/resources`, where "#" is the API
version number.

#### Navigating API Versions
Leaving `/v#/` out of your API query URLs will default to the most recent,
stable API verison. To prevent unexpected results, it is advised to always
supply a version number.

### Syntax
#### The Query String
##### Special Character Meanings
- Forward-slash, `/`: Only for URL pathing.
- Back-slash, `\ `: For character escaping (TODO).
- Question mark, `?`: For specifying query parameters.
- Ampersand, `&`: Query parameter delimiter.
- Equals-sign, `=`: Keyword query parameter assignment operator.
- Comma, `,`: List item delimiter.

##### Character Escpaing
TODO

#### Data Types
##### Primitives
In an API query URL string, specific data types should be used as follows and
should be specified witout quotation marks (`"` or `'`).
###### String
Example: `/v1/datalab/data?survey=GH2013PMA`.

###### Integer
Example: `/v1/datalab/data?precision=1`.

###### Float
Example: `/v1/datalab/data?value=15.4`.

###### Boolean
Booleans can be any upper or lower case variant of 'true' or 'false'. Values
"1", "0", "null", "none" are examples of non-booleans.
Example: `/v1/datalab/indicator?isFavorite=true`.

###### Null
Null can be any upper or lower case variant of 'null'.  Values "0" and "none"
are examples of non-null values.
Example: `/v1/datalab/indicator?isFavorite=null`.

###### None
None can be any upper or lower case variant of 'none'.  Values "0" and "null"
are examples of non-none values.
Example: `/v1/data?char1.id=none`

##### Lists
Lists are simply comma separated strings without qotation marks (`"` or `'`).
Example: `/v1/datalab/combos?survey=GH2013PMA,GH2014PMA`

#### Querying Collections and Entities
TODO

#### Query Parameters
##### Filtering
TODO

##### Fields
TODO

##### Format
TODO


### Common Endpoints

#### Root `/v1/`
This endpoint re-routes directly to: `/v1/resources`

#### Resources `/v1/resources`
Returns a list of core resources. Takes no arguments.

Example: `/v1/resources`
```
{
  "resources": [
    {
      "name": "countries",
      "resource": "https://<base_url>/surveys"
    },
    {
      "name": "surveys",
      "resource": "https://<base_url>/surveys"
    },
    {
      "name": "texts",
      "resource": "https://<base_url>/texts"
    },
    {
      "name": "indicators",
      "resource": "https://<base_url>/indicators"
    },
    {
      "name": "data",
      "resource": "https://<base_url>/data"
    },
    {
      "name": "characteristicGroups",
      "resource": "https://<base_url>/characteristicGroups"
    }
  ]
}
```

#### Version `/v1/version`
Returns API 2-part semantic version number of form "#.#". Takes no arguments.

Example: `/v1/version`
```
{
    "version": "0.1"
}
```

### Application Specific Endpoints
This is documentation for application specific endpoints, such as
[PMA2020 Datalab](http://datalab.pma2020.org).

#### Application initialization `/v1/datalab/init`
Returns lists of all specific, key resources (surveys, indicators, and
characteristics), all of which have at least one stored data point associated.
Takes no arguments.

Example: `/v1/datalab/init`
```
{
  "characteristicGroups": [
    {
      "category.id": "To be implemented.",
      "definition.id": "0UI5Qdrh",
      "id": "wealth_quintile",
      "label.id": "BP8WlFPW",
      "order": "To be implemented."
    },
    ...
  ],
  "characteristics": [
    {
      "id": "none",
      "label.id": "jVwfpK1a",
      "order": 0
    },
    ...
  ],
  "indicators": [
    {
      "category.id": "h2koO9c-",
      "definition.id": "e3E-GqDL",
      "id": "uneed_tot_aw",
      "label.id": "e3E-GqDL",
      "order": 80
    },
    ...
  ],
  "languages": {
    "en": "English",
    "fr": "French",
    ...
  },
  "strings": {
    "-iTSAPqU": {
      "en": "Non-user"
    },
    "-lfInUMt": {
      "en": "Highest"
    },
    ...
  },
  "surveys": [
    {
      "country.label.id": 3,
      "geography.label.id": "To be implemented.",
      "id": "KE2014PMA",
      "label.id": "To be implemented.",
      "order": 601
    },
    ...
  ]
}
```

#### Query for valid combinations of key resources `/v1/datalab/combos`
Query for valid combinations of key resources, given any combination of 3 possible key
resources required for rendering visualizations. A valid combinations of
resources is defined as resources that, when passed to `v1/data` or
`v1/datalab/data`, will return 1 or more data points. This endpoint requires
1-3 of the following parameters (key resources): "survey", "indicator", or
"characteristicGroup".

##### A. Query by survey `/v1/datalab/combos?survey=<surveys>`
A list of indicator and characteristic combinations which have one or more
stored data point associated with any of the provided surveys. From this list
provided, the client application should have enough information to filter valid
data isaggregated further by specified indicator(s) and characteristic(s),
without the need for any further API calls.

Example: `/v1/datalab/combos?survey=GH2013PMA,GH2014PMA`
```
{
  "characteristicGroup.id": [
    "none",
    "residence",
    "wealth_quintile"
  ],
  "indicator.id": [
    "mcpr_aw",
    "uneed_tot_aw"
  ],
  "survey.id": [
    "GH2013PMA",
    "GH2014PMA",
    "GH2015PMA",
    "KE2014PMA",
    "KE2015PMA",
    "KE2015PMA_2"
  ]
}
```

##### B. Query by indicator or characteristicGroup `/v1/datalab?[indicator | characteristicGroup]=<id>`
Filtering by either indicator or characteristicGroup returns a list of
indicators or characteristics (whichever was not provided in the query) which
have one or more stored data point associated, and a separate list of surveys
which have one or more stored data point associated.

Example: `/v1/datalab/combos?indicator=mcpr_aw`
```
{
  "characteristicGroup.id": [
    "none",
    "residence",
    "wealth_quintile"
  ],
  "indicator.id": [
    "mcpr_aw",
    "uneed_tot_aw"
  ],
  "survey.id": [
    "GH2013PMA",
    "GH2014PMA",
    "GH2015PMA",
    "KE2014PMA",
    "KE2015PMA",
    "KE2015PMA_2"
  ]
}
```

Example: `/v1/datalab/combos?characteristicGroup=wealth_quintile`
```
{
  "characteristicGroup.id": [
    "none",
    "residence",
    "wealth_quintile"
  ],
  "indicator.id": [
    "mcpr_aw",
    "uneed_tot_aw"
  ],
  "survey.id": [
    "GH2013PMA",
    "GH2014PMA",
    "GH2015PMA",
    "KE2014PMA",
    "KE2015PMA",
    "KE2015PMA_2"
  ]
}
```

##### C. Query by 2 of 3 key resources `/v1/datalab/combos?<resource_1>=<id>&<resource_2>=<id>`
Filter by 2 of 3 of the `/datalab/combos` parameters: "survey", "indicator", and
"characteristicGroup".

Example: `/v1/datalab/combos?survey=GH2015PMA&indicator=mcpr_aw`
```
{
  "characteristicGroup.id": [
    "none",
    "residence",
    "wealth_quintile"
  ],
  "indicator.id": [
    "mcpr_aw",
    "uneed_tot_aw"
  ],
  "survey.id": [
    "GH2013PMA",
    "GH2014PMA",
    "GH2015PMA",
    "KE2014PMA",
    "KE2015PMA",
    "KE2015PMA_2"
  ]
}
```

#### Querying application-specific data `/v1/datalab/data`
Returns list of data points with minimal number of fields necessary for
rendering visualizations.

##### Required Parameters
- `survey` *(string)* - Survey.
- `indicator` *(string)* - Indicator.
- `characteristicGroup` *(string)* - Characteristic group.

Example: `/v1/datalab/data?survey=GH2013PMA,GH2014PMA&indicator=mcpr_aw&characteristicGroup=none`
```
{
  "chartOptions": {
    "precision": 1
  },
  "metadata": {
    "datasetMetadata": [
      {
        "createdOn": "Thu, 14 Sep 2017 20:32:02 GMT",
        "hash": "ef5c8634e0dbc812a3df9b0882873db9",
        "name": "api_data",
        "type": "api"
      },
      {
        "createdOn": "Thu, 14 Sep 2017 20:32:02 GMT",
        "hash": "cca82c783607b2bf1431ac5bcd26f8cc",
        "name": "ui_data",
        "type": "ui"
      }
    ],
    "version": "0.1.8"
  },
  "queryInput": {
    "characteristicGroups": [
      {
        "definition.id": "PRE_iU_n",
        "id": "none",
        "label.id": "XkVmGDDF"
      }
    ],
    "indicators": [
      {
        "definition.id": "NQv8ZZOp",
        "id": "mcpr_aw",
        "label.id": "NQv8ZZOp"
      }
    ],
    "surveys": [
      {
        "country.label.id": "6EA0At85",
        "geography.label.id": "w748V1ul",
        "id": "GH2013PMA",
        "label.id": "lq0db_sX",
        "partner.label.id": "cizmJ6Gv"
      },
      {
        "country.label.id": "6EA0At85",
        "geography.label.id": "w748V1ul",
        "id": "GH2014PMA",
        "label.id": "2Ea5SlF4",
        "partner.label.id": "cizmJ6Gv"
      }
    ]
  },
  "resultSize": 2,
  "results": [
    {
      "country.id": "GH",
      "country.label.id": "6EA0At85",
      "geography.id": "gh_national",
      "geography.label.id": "w748V1ul",
      "survey.id": "GH2013PMA",
      "survey.label.id": "lq0db_sX",
      "values": [
        {
          "characteristic.id": "none",
          "characteristic.label.id": "J-N_aTkS",
          "value": 15.4
        }
      ]
    },
    {
      "country.id": "GH",
      "country.label.id": "6EA0At85",
      "geography.id": "gh_national",
      "geography.label.id": "w748V1ul",
      "survey.id": "GH2014PMA",
      "survey.label.id": "2Ea5SlF4",
      "values": [
        {
          "characteristic.id": "none",
          "characteristic.label.id": "J-N_aTkS",
          "value": 16.1
        }
      ]
    }
  ]
}
```

##### Optional Parameters
###### Querying Time Series Data
- `overTime` *(boolean)*

Supplying this query parameter and setting value to `true` will tell the API
to return chronologically sorted data, and also include a `date` attribute for
 each data point. If not supplied, the API will automatically interpret this
 value as `false`.

Example: `/v1/datalab/data?survey=GH2013PMA,GH2014PMA&indicator=mcpr_aw&characteristicGroup=none&overTime=true`
```
{
  "chartOptions": {
    "precision": 1
  },
  "metadata": {
    "datasetMetadata": [
      {
        "createdOn": "Thu, 14 Sep 2017 20:32:02 GMT",
        "hash": "ef5c8634e0dbc812a3df9b0882873db9",
        "name": "api_data",
        "type": "api"
      },
      {
        "createdOn": "Thu, 14 Sep 2017 20:32:02 GMT",
        "hash": "cca82c783607b2bf1431ac5bcd26f8cc",
        "name": "ui_data",
        "type": "ui"
      }
    ],
    "version": "0.1.8"
  },
  "queryInput": {
    "characteristicGroups": [
      {
        "definition.id": "PRE_iU_n",
        "id": "none",
        "label.id": "XkVmGDDF"
      }
    ],
    "indicators": [
      {
        "definition.id": "NQv8ZZOp",
        "id": "mcpr_aw",
        "label.id": "NQv8ZZOp"
      }
    ],
    "surveys": [
      {
        "country.label.id": "6EA0At85",
        "geography.label.id": "w748V1ul",
        "id": "GH2013PMA",
        "label.id": "lq0db_sX",
        "partner.label.id": "cizmJ6Gv"
      },
      {
        "country.label.id": "6EA0At85",
        "geography.label.id": "w748V1ul",
        "id": "GH2014PMA",
        "label.id": "2Ea5SlF4",
        "partner.label.id": "cizmJ6Gv"
      }
    ]
  },
  "resultSize": 1,
  "results": [
    {
      "characteristic.id": "none",
      "characteristic.label.id": "J-N_aTkS",
      "country.id": "GH",
      "country.label.id": "6EA0At85",
      "geography.id": "gh_national",
      "geography.label.id": "w748V1ul",
      "values": [
        {
          "survey.date": "2013-10-03",
          "survey.id": "GH2013PMA",
          "survey.label.id": "lq0db_sX",
          "value": 15.4
        },
        {
          "survey.date": "2014-01-01",
          "survey.id": "GH2014PMA",
          "survey.label.id": "2Ea5SlF4",
          "value": 16.1
        }
      ]
    }
  ]
}
```
###### CSV download
- `format=csv`
Add `format=csv` to the query parameters to get a csv download.
- `lang=LC`
Add `lang=LC` where `LC` is the language code (e.g. `en` or `fr`) for the data download. This is optional, and the default is English.

## 2.2 Installation
### System dependencies
- [Python3](https://www.python.org/downloads/)
- SQLite: There are several options. We reccomend [this](https://www.sqlite.org/download.html) for CLI, or [this](https://sqlitebrowser.org/) for GUI.

### Clone the project
1. Change directory to `PATH` where you want to clone: `cd PATH`
2. Clone: `git clone https://github.com/PMA-2020/pma-api.git`

### Virtual environment _(optional, but recommended)_
A Python virtual environment is a way to isolate applications and their dependencies from one another in order to avoid conflicts. We recommend that you install virtualenv. [The virtualenv documentation](https://virtualenv.pypa.io/en/stable/) has installation steps as well as more information about the usefulness of virtual environments. Here's a summary of the installation steps:

1. Install `virtualenv` globally: `python3 -m pip install virtualenv`
2. Change directory to where `pma-api` was cloned: `cd PATH/pma-api`
3. Create a virtual environment called `env`: `virtualenv env`

### Install project dependencies
- `python3 -m pip install -r requirements.txt`

### Create the database _(optional)_
1. Create the database via the following `makefile` command: `make db`
_This will execute a script that populates a database using data taken from the `data` directory. In a properly set up staging or production environment, PostgreSQL will be used. In a development environment, SQLite is used. This step is considered optional in a development environment as there should already be an up-to-date `dev.db` in the root directory._
2. Examine database to verify that all looks well. _(optional)_
This can be done via CLI or GUI tool, such as recommended previously.

## 2.3 Using locally
1. Run `pma-api` on a local server process via the following makefile command: `make serve`
2. Verify that it is running in the browser by going to: `http://localhost:5000/v1/resources`
3. Navigate the API further by either (a) utilizing the URL links shown in the browser, or (b) looking at the available endpoints in the [pma-api documentation](https://www.github.com/PMA-2020/pma-api).
4. Read the [pma-api documentation](https://www.github.com/PMA-2020/pma-api) for more information on use.

## 2.4 Server management
### Pushing to staging
`make push-staging`

### Pushing to production
`make push-production`

CHANGES.md
# v1.1.0, 25 July 2018
## New features
- Added dedicated docs, accessible at http://api-docs.pma2020.org

## General updates
- Default route now redirects to docs.
- Versioned to 1.1.0, "1.x" to keep consistent with the API route version 'v1', and "1.1.x" to not cause any confusion that this is "the initial release".

## Developer updates
- Added continuous integration
- Changed changelog to Markdown format.

# v0.1.9, 14 November 2017
- Group /datalab/data over time correctly if only one characteristic
- Better, more readable CSV downloads with other languages supported

# v0.1.8, 10 October 2017
- Fix characteristic group bug in /datalab/data
- Move precision to top level under 'chartOptions'
- All values returned after rounding
- Dataset update

# v0.1.7, 2 October 2017
- Empty CSV download is now a 204 server response
- Implemented server-side caching for /datalab/init
- Added query param cached=false to force dynamic response
- Various dataset fixes, including date MM-YYYY

# v0.1.6, 14 September 2017
- Send back datalab/init data for query parameters to /datalab/data
- Update documentation with overTime and format=csv

# v0.1.5, 13 September 2017
- Echo back query parameters for /datalab/data and
   /datalab/combos
- CSV download fixed
- Support /datalab/data?overTime=true
- Add in partner information with survey
- Change /datalab/combos so that it always returns three
   lists of what is valid to choose for the datalab

# v0.1.4, 11 September 2017
- Small documentation update
- 'label.id' used in /datalab/init result
- /datalab/data returns a list of series
- CSV format broken for now

# v0.1.3, 6 September 2017
- CSV format allowed on /v1/datalab/data?format=csv
- Support CORS
- Version information moved to /version
- Internal refactor to return custom object

# v0.1.2, 1 September 2017
- Datalab init for indicators fixed
- Added back '/' root route
- Added support for UI translations

# v0.1.1, 31 August 2017
- `/v1/version` route and JSON objects return metadata object
   which shows version information for code and source data files.
- datalab/init has better-formatted objects for indicator,
   characteristicGroup, and country-round select boxes.
- Added metadata table to database

pma_api/__init__.py
"""Definition of application object."""
import os

from flask import Blueprint, jsonify, redirect, request
from flask_cors import CORS

from .app import PmaApiFlask
from .config import config
from .models import db
from .response import QuerySetApiResult


root = Blueprint('root', __name__)


@root.route('/')
def root_route():
    """Root route.

    .. :quickref: Root; Redirects to resources list or documentation depending
     on MIME type

    Args:
        n/a

    Returns:
        func: get_resources() if 'application/json'
        func: redirect() to docs if 'text/html'

    Details:
        Redirects implicitly if MIME type does not explicitly match what is
        expected.
    """
    request_headers = request.accept_mimetypes\
        .best_match(['application/json', 'text/html'])

    if request_headers == 'text/html':
        return redirect('http://api-docs.pma2020.org', code=302)
    else:
        from .api_1_0.collection import get_resources as res
        return res() if request_headers == 'application/json' else res()


@root.route('/docs')
def documentation():
    """Documentation.

    .. :quickref: Docs; Redirects to official documentation.

    Args:
        n/a

    Returns:
        redirect(): Redirects to official documentation.
    """
    return redirect('http://api-docs.pma2020.org', code=302)


@root.route('/version')
def show_version():
    """Show API version data.

    .. :quickref: Version; API versioning data.

    Args:
        n/a

    Returns:
        String: Version number.

    Details:
        Displays API's 2-part semantic version number. ALso displays versioning
        for datasets used.

    Example:
        .. code-block:: json
           :caption: GET /v1/version
           :name: example-of-endpoint-version

            {
              "datasetMetadata": [
                {
                  "createdOn": "Fri, 13 Jul 2018 20:25:42 GMT",
                  "hash": "339ce036bdee399d449f95a1d4b3bb8f",
                  "name": "api_data-2018.03.19-v29-SAS",
                  "type": "api"
                },
                {
                  "createdOn": "Fri, 13 Jul 2018 20:25:43 GMT",
                  "hash": "469542a93241da0af80269b6d7352600",
                  "name": "ui_data-2017.10.02-v4-jef",
                  "type": "ui"
                }
              ],
              "version": "0.1.9"
            }
    """
    return jsonify(QuerySetApiResult.metadata())


def create_app(config_name=os.getenv('FLASK_CONFIG', 'default')):
    """Create configured Flask application.

    Args:
        config_name (str): Name of the configuration to be used.

    Returns:
        Flask: Configured Flask application.
    """
    # noinspection PyShadowingNames
    app = PmaApiFlask(__name__)
    app.config.from_object(config[config_name])

    CORS(app)
    db.init_app(app)
    app.register_blueprint(root)

    from .api_1_0 import api as api_1_0_blueprint
    app.register_blueprint(api_1_0_blueprint, url_prefix='/v1')

    return app
